### PR TITLE
Universum 1.0.0

### LINKED ISSUE #5: Figure out what to do with unsafe `head` and friends
Currently:
- `Unsafe` exports `unsafe(Head|Tail|Init|Last|At)`
- `head` returns `Maybe`
- `tail`, `init`, `last`, `(!!)` aren't exported at all
- functions from safe (e.g. `atMay`, `atDef`, `headMay`, `headDef`) are exported by default

I don't have a strong preference here – the only thing I kinda want to do is to remove `head`. Otherwise I'm slightly in favor of removing `unsafe*` variants and adding unsafe `head`, etc. back while retaining safe variants, but I'm okay with not doing it. What do others think?

### LINKED ISSUE #56: 'elem' is O(n) for sets :(
Sets usually have O(log n) `elem`, but IntSet and HashSet have O(n) `elem` because the `elem` from `Foldable` is slower than it could be (since it can't use the `Hashable` or `Ord` constraints). Facepalm.

We can fix it by doing something like this, as David Feuer proposed at libraries@haskell.org:

```haskell
class Foldable f where
  type ElemConstr t :: * -> Constraint
  type ElemConstr t = Eq

  elem :: ElemConstr t a => a -> t a -> Bool
  default elem :: (ElemConstr t ~ Eq, Eq a) => a -> t a -> Bool
```

(But for `NontrivialContainer` instead of `Foldable`.)

### LINKED ISSUE #58: Make 'Element' an associated type family
It was suggested under comments to [this answer on SO](https://stackoverflow.com/questions/45424976/foldable-intset/45425178#comment77841903_45425178).

### LINKED ISSUE #81: `putText` and `putTextLn`
For the moment there is only `putText` function in `universum` which behaves like `putStrLn`. 
I think it would be convenient to have `putText` which would be specialized version of `putStr` and `putTextLn` — version of `putStrLn`.

### LINKED ISSUE #88: Add 'HasCallStack =>' to 'error' and 'undefined'
Here:

https://github.com/serokell/universum/blob/5ba8834ad89d0cc1e7fb537eb7dc9c92162e3b49/src/Debug.hs#L42

Then the stack will give something more useful.

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #5 
Resolves #56
Resolves #58
Resolves #81 
Resolves #88 

Also `unsnoc` is removed.
`safe` package is removed from dependencies.

### CODE REVIEW

File: src/Universum/Container/Class.hs [hunk 1]
```diff
@@ -114,7 +101,12 @@ type instance Element IS.IntSet = Int
 -- @'null' ≡ 'List.null' . 'toList'@
 --
 class ToList t where
-    {-# MINIMAL toList #-}
+    -- | Type of element for some container. Implemented as a type family because
+    -- some containers are monomorphic over element type (like 'T.Text', 'IS.IntSet', etc.)
+    -- so we can't implement nice interface using old higher-kinded types approach.
+    type Element t :: *
```
Comment (vrom911) [inline src/Universum/Container/Class.hs]: Maybe add explanation into docs why it has to be inside `ToList` class?

**Comment by Anton-Latukha** [comment]
🕋

File: src/Universum/Container/Class.hs [hunk 2]
```diff
@@ -42,28 +39,27 @@ module Universum.Container.Class
...
-import Data.Maybe (fromMaybe)
-import Data.Monoid (All (..), Any (..), First (..))
-import Data.Word (Word8)
-import Prelude hiding (Foldable (..), all, and, any, head, mapM_, notElem, or, sequence_)
+import Prelude hiding (all, and, any, elem, foldMap, foldl, foldr, mapM_, notElem, or, product,
+                sequence_, sum)
+
+import Universum.Applicative (Alternative (..), Const, ZipList, pass)
+import Universum.Base (Constraint, Foldable, Word8)
+import Universum.Container.Reexport
```
[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Universum/Container/Class.hs]: I agree. And I just noticed that not everything from this module is used. That's why it's better to make this import explicit.

File: src/Universum/Container/Class.hs (hunk 1 above)
Comment (ChShersh) [reply src/Universum/Container/Class.hs]: I added some explanation. But I don't know how to explain benefits because it's hard to explain...

**Pull request merged**